File: GGAndroid_2.0/app/src/main/java/io/ganguo/incubator/viewmodel/activity/ApkInfoViewModel.java
package io.ganguo.incubator.viewmodel.activity;

import android.os.Build;
import android.view.View;
import android.view.ViewGroup;

import io.ganguo.incubator.BuildConfig;
import io.ganguo.incubator.R;
import io.ganguo.library.ui.view.ActivityInterface;
import io.ganguo.utils.common.ResHelper;
import io.ganguo.viewmodel.common.HFRecyclerViewModel;
import io.ganguo.viewmodel.common.HeaderItemViewModel;
import io.ganguo.viewmodel.common.HeaderViewModel;
import io.ganguo.viewmodel.common.TextViewModel;
import io.ganguo.viewmodel.databinding.IncludeHfRecyclerBinding;
import io.ganguo.vmodel.BaseViewModel;
import io.ganguo.vmodel.ViewModelHelper;

/**
 * Apk相关数据信息
 * Created by leo on 2018/11/1.
 */
public class ApkInfoViewModel extends HFRecyclerViewModel<ActivityInterface<IncludeHfRecyclerBinding>> {
    @Override
    public void initHeader(ViewGroup container) {
        super.initHeader(container);
        ViewModelHelper.bind(container, this, createHeaderVModel());
    }

    @Override
    public void onViewAttached(View view) {
        addApkViewModel();
    }

    /**
     * function：add apk info
     */
    protected void addApkViewModel() {
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_package_info)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_packaging_environment, BuildConfig.ENVIRONMENT)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_packaging_base_url, BuildConfig.BASE_URL)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_packaging_time, BuildConfig.PACKAGING_TIME)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_version_sdk, Build.VERSION.SDK_INT)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_sys_version, Build.VERSION.RELEASE)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_device_model, Build.MODEL)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_version_name, BuildConfig.VERSION_NAME)));
        getAdapter().add(createTextVModel(ResHelper.getString(R.string.str_version_code, BuildConfig.VERSION_CODE)));
        getAdapter().notifyDataSetChanged();
        toggleEmptyView();
    }


    /**
     * function：create header ViewModel
     *
     * @return
     */
    protected BaseViewModel createHeaderVModel() {
        return new HeaderViewModel.Builder()
                .appendItemLeft(new HeaderItemViewModel
                        .BackItemViewModel(getView().getActivity())
                        .paddingRes(R.dimen.dp_16)
                        .drawableRes(android.R.drawable.ic_menu_close_clear_cancel))
                .appendItemCenter(new HeaderItemViewModel
                        .TitleItemViewModel(getStrings(R.string.str_packaging_apk_info)))
                .appendItemLeft(new HeaderItemViewModel
                        .BackItemViewModel(getView().getActivity()))
                .build();
    }


    /**
     * function: create TextViewModel
     *
     * @param text
     * @return
     */
    private BaseViewModel createTextVModel(String text) {
        return new TextViewModel
                .Builder()
                .content(text)
                .textSizeRes(R.dimen.font_16)
                .paddingRightRes(R.dimen.dp_20)
                .paddingLeftRes(R.dimen.dp_20)
                .textColorRes(R.color.black)
                .width(ViewGroup.LayoutParams.MATCH_PARENT)
                .paddingBottomRes(R.dimen.dp_10)
                .paddingTopRes(R.dimen.dp_10)
                .build();
    }
}

File: GGAndroid_3.0/main/app/src/main/java/io/ganguo/incubator/viewmodel/general/activity/ApkInfoViewModel.kt
package io.ganguo.incubator.viewmodel.general.activity

import android.os.Build
import android.view.View
import android.view.ViewGroup
import io.ganguo.app.core.lifecycle.ApkInfoPageObserver
import io.ganguo.incubator.AppEnv
import io.ganguo.incubator.R
import io.ganguo.viewmodel.core.viewinterface.ActivityInterface
import io.ganguo.utils.helper.ResHelper
import io.ganguo.viewmodel.core.BaseViewModel
import io.ganguo.viewmodel.core.ViewModelHelper
import io.ganguo.viewmodel.databinding.IncludeHfRecyclerBinding
import io.ganguo.viewmodel.pack.base.viewmodel.BaseHFRecyclerVModel
import io.ganguo.viewmodel.pack.common.HeaderItemViewModel
import io.ganguo.viewmodel.pack.common.HeaderViewModel
import io.ganguo.viewmodel.pack.common.TextViewModel

/**
 * <pre>
 * author : leo
 * time   : 2018/9/15
 * desc   : Apk及手机相关参数展示 ViewModel
 * </pre>
 *
 */
class ApkInfoViewModel : BaseHFRecyclerVModel<ActivityInterface<IncludeHfRecyclerBinding>>() {

    init {
        addObserver()
    }

    override fun initHeader(container: ViewGroup) {
        super.initHeader(container)
        ViewModelHelper.bind(container, this, createHeaderVModel())
    }

    override fun onViewAttached(view: View) {
        addApkViewModel()
    }


    /**
     * 添加页面生命周期观察类
     */
    private fun addObserver() {
        lifecycle.addObserver(ApkInfoPageObserver(lifecycle) {
            //todo: refresh data
        })
    }

    /**
     * function：add apk info
     */
    private fun addApkViewModel() {
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_package_info)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_packaging_time, AppEnv.packagingTime)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_package_name, AppEnv.packageName)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_version_name, AppEnv.versionName)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_version_code, AppEnv.versionCode)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_packaging_environment, AppEnv.environment)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_packaging_base_url, AppEnv.baseUrl)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_version_sdk, Build.VERSION.SDK_INT)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_sys_version, Build.VERSION.RELEASE)))
        adapter.add(createTextVModel(ResHelper.getString(R.string.str_app_device_model, Build.MODEL)))
        adapter.notifyDataSetChanged()
        toggleEmptyView()
    }


    /**
     * function：create header ViewModel
     *
     * @return
     */
    private fun createHeaderVModel(): BaseViewModel<*> {
        return HeaderViewModel.Builder()
                .appendItemLeft(HeaderItemViewModel.BackItemViewModel(viewInterface.activity)
                        .paddingRes(R.dimen.dp_16)
                        .drawableRes(android.R.drawable.ic_menu_close_clear_cancel))
                .appendItemCenter(HeaderItemViewModel.TitleItemViewModel(getString(R.string.str_app_packaging_apk_info)))
                .appendItemLeft(HeaderItemViewModel.BackItemViewModel(viewInterface.activity))
                .build()
    }


    /**
     * function: create TextViewModel
     *
     * @param text
     * @return
     */
    private fun createTextVModel(text: String): BaseViewModel<*> {
        return TextViewModel.Builder()
                .content(text)
                .textSizeRes(R.dimen.font_16)
                .paddingRightRes(R.dimen.dp_20)
                .paddingLeftRes(R.dimen.dp_20)
                .textColorRes(R.color.black)
                .width(ViewGroup.LayoutParams.MATCH_PARENT)
                .paddingBottomRes(R.dimen.dp_10)
                .paddingTopRes(R.dimen.dp_10)
                .build()
    }
}
